### DISCUSSION TITLE
Compiling an external plugin

### DISCUSSION THREAD

**Discussion by wz2b** [strongswan/strongswan#2461] [Q&A]
I'm on an Ubuntu system and I installed strongswan, which includes the libstrongswan package, version 5.9.13-ubuntu4.  This got me /usr/lib/ipsec/libstrongswan.so.0.0.0 but it did not get me a strongswan.h anywhere in /usr/include.  Is it still possible to build a charon plugin completely externally?  I downloaded the source tarball, configured and built that successfully, but that doesn't include a strongswan.h either.

**Comment by tobiasbrunner** [comment]
Where did you ever see a `strongswan.h`?

While there is the configure option `--with-dev-headers`, this only installs the libstrongswan, libtls and libfast headers somewhere in the system (it also puts `config.h` in `ipseclibdir`, which is necessary to compile anything with those headers). Since we don't provide a stable API, it's not the intention that distributions enable this option (it was added for some internal projects at the time, not for general use).

Anyway, the main header of libstrongswan is `library.h`. However, depending on the type of plugin you intend to develop, those headers won't be enough and you're better off referencing the library directories in the source dir. You'll still need a generated `config.h` that matches the configure options/results of the installed binaries, so ideally you get it from a build that uses the same options and toolchain used by Ubuntu.

See [this example out-of-tree plugin](https://github.com/tobiasbrunner/strongswan-redirect-plugin) for some additional pointers.

**Reply by wz2b** [reply]
I saw strongswan.h in some search results that I'm now thinking were AI generated and are quite suspect.  Thanks very much for pointing me to that example.
